REGRESSION (258514@main): Transition of !important property fails to animate

anttijk opened the pull request:
#### e0fa0f2073f1886a7272a69040adaaafde9754fd
<pre>
REGRESSION (258514@main): Transition of !important property fails to animate
<a href="https://bugs.webkit.org/show_bug.cgi?id=111329">https://bugs.webkit.org/show_bug.cgi?id=111329</a>
rdar://105929421

Reviewed by Antoine Quint.

The cascade precedence is (<a href="https://drafts.csswg.org/css-cascade-5/#cascade-sort)">https://drafts.csswg.org/css-cascade-5/#cascade-sort)</a>

1) Transition declarations [css-transitions-1]
2) Important user agent declarations
3) Important user declarations
4) Important author declarations
5) Animation declarations [css-animations-1]
6) Normal author declarations
7) Normal user declarations
8) Normal user agent declarations

but we were treating transitions and animations similarly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-important-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-important.html: Added.
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch):
(WebCore::Style::PropertyCascade::shouldApplyAfterAnimation):

Don&apos;t let !important properties override transitions.

* Source/WebCore/style/PropertyCascade.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Tell the cascade if this is a transition or animation.
Note that this is still incorrect in the (rare) case where you would have both running on an element at the same time.

(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/260880@main">https://commits.webkit.org/260880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/54738d139f5a4da8ea0ae1476dee14ec33bd827a

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109775 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18890 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1245 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118895 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113727 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20361 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10088 "Built successfully") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/102065 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/15159 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98387 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43381 "layout-tests (failure)") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/97126 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/85178 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11616 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31382 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12268 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8325 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17637 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50991 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7547 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/14018 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-important.html:
```diff
@@ -0,0 +1,24 @@
+test(t => {
+    document.body.offsetLeft;
+    target.className = "red";
+    assert_equals(getComputedStyle(target).backgroundColor, "rgb(0, 128, 0)");
```
While it's functionally equivalent, I think it would be better form to query the computed style here instead of getting `offsetLeft` to ensure a style update prior to apply transitions.

```suggestion
    const cs = getComputedStyle(target);
    target.className = "red";
    assert_equals(cs.backgroundColor, "rgb(0, 128, 0)");
```

graouts commented on LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-important.html:
```diff
@@ -0,0 +1,24 @@
+    width: 200px;
+    height: 200px;
+    background-color: green;
+    transition: background-color 1s;
```
I think it might be best to make the transition duration super long to ensure we are as close as possible to a `0` progress value when we query the computed style after enforcing the transition. The test might be super stable as it is, but using something like `1000s` would not hurt.

graouts commented on Source/WebCore/style/StyleTreeResolver.cpp:
```diff
@@ -611,7 +611,8 @@ ElementUpdate TreeResolver::createAnimatedElementUpdate(ResolvedStyle&& resolved
             auto animatedStyleBeforeCascadeApplication = RenderStyle::clonePtr(*animatedStyle);
             // The cascade may override animated properties and have dependencies to them.
-            applyCascadeAfterAnimation(*animatedStyle, animatedProperties, *resolvedStyle.matchResult, element, resolutionContext);
+            // FIXME: This is wrong if there are both transitions and animations running on the same element.
```
Would have been nice to write a test where we fail because of this specific case and – at a minimum – file a bug for us to address this.

anttijk replied:
https://bugs.webkit.org/show_bug.cgi?id=252985

anttijk committed:
REGRESSION (258514@main): Transition of !important property fails to animate
https://bugs.webkit.org/show_bug.cgi?id=111329
rdar://105929421

Reviewed by Antoine Quint.

The cascade precedence is (https://drafts.csswg.org/css-cascade-5/#cascade-sort)

1) Transition declarations [css-transitions-1]
2) Important user agent declarations
3) Important user declarations
4) Important author declarations
5) Animation declarations [css-animations-1]
6) Normal author declarations
7) Normal user declarations
8) Normal user agent declarations

but we were treating transitions and animations similarly.

* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-important-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-transitions/transition-important.html: Added.
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch):
(WebCore::Style::PropertyCascade::shouldApplyAfterAnimation):

Don't let !important properties override transitions.

* Source/WebCore/style/PropertyCascade.h:
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):

Tell the cascade if this is a transition or animation.
Note that this is still incorrect in the (rare) case where you would have both running on an element at the same time.

(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/260880@main
